### PR TITLE
Write tests for invertedIndex class, Implement the class, code coverage and coveralls, and travis CI

### PR DESCRIPTION
**Pull request by desaawa**
#### What does this PR do?
Merges jasmine test and an implementation of invertedIndex class
#### Description of Task to be completed?
Write all the tests for the application and create an inverted index class
#### How should this be manually tested?
Clone the repo and install locally.
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
N/A
#### Screenshots (if appropriate)
N/A
#### Questions:PR Naming
N/A

### CODE REVIEW

File: jasmine/spec/inverted-index-test.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    });
+
+    it('filters out symbols', () => {
+      expect(InvertedIndex.getWords('alice # in* Won@derland'))
+        .toEqual(['alice', 'in', 'wonderland']);
+    });
+  });
+
+  describe('CreateIndex', () => {
+    it('creates an index', () => {
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js:37]: Question: Is the index created from the title or the test of each document?

Also, I do not see you calling the `createIndex` method here where you're testing the result.

File: jasmine/spec/inverted-index-test.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
+const books = require('./books.json')
+
+
+describe('InvertedIndex Class', () => {
+  beforeEach(() => {
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js]: What's the implication of `beforeEach` as compared to `beforeAll`? Do you think you need to call these methods before each of the specs?

File: jasmine/spec/inverted-index-test.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    it('filters out symbols', () => {
+      expect(InvertedIndex.getWords('alice # in* Won@derland'))
+        .toEqual(['alice', 'in', 'wonderland']);
+    });
+  });
+
+  describe('CreateIndex', () => {
+    it('creates an index', () => {
+      expect(this.index.a).toEqual([0, 1, 2]);
+      expect(this.index.alice).toEqual([0]);
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js]: Does this guarantee that Index has been created and that the Index is correct?

File: jasmine/spec/inverted-index-test.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    });
+  });
+
+
+  describe('GetIndex', () => {
+    it('should return a particular index', () => {
+      const index = this.invertedIndex.getIndex('books');
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js]: You have just one index in the entire test. How do we know the function would get the correct index for each of them.

Try to test that and also test the exact result of the index, not just for a few words in here.

Also, you should test that an index is not returned when no book with said name was uploaded.

File: jasmine/spec/inverted-index-test.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    });
+
+    it('should return "not found" for words not in index', () => {
+      expect(this.invertedIndex.searchIndex('', 'books'))
+        .toEqual('no word found');
+    });
+
+    it('should return object with search words', () => {
+      const result = this.invertedIndex
+                         .searchIndex('alice in wonderland', 'books');
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js]: From your test it's unclear that the search result will return the accurate documents in which these words occur. Please let's ensure to do that.

File: jasmine/spec/app-test.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
+(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
```
Comment (q-ode) [inline jasmine/spec/app-test.js]: If this is a generated file which I believe it is, it shouldn't be in this folder and definitely shouldn't be committed to the repo. You should run your tasks during deployment to have them generated on the server.

File: jasmine/spec/inverted-index-test.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
+const books = require('./books.json')
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js]: You're missing a semicolon here.

Is your HoundCI working? I surprised it's not flagging this.

**Review by q-ode — COMMENTED** [review]
Great Job! Overall.

I've left comments specific to your tests, kindly address them.

File: jasmine/spec/inverted-index-test.js (hunk 7 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js]: I have added the semicolon. Thanks

I think my HoundCI is working.

File: jasmine/spec/inverted-index-test.js (hunk 2 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js]: Thanks. I think `beforeAll` will be better, since I will not be calling the methods before each and every spec. With `beforeAll` I will initialize the class and create the indexes once and for all, without repeating myself before each spec.

File: jasmine/spec/inverted-index-test.js (hunk 1 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js:37]: The Index is created from both title and text. I've tested for both.

I already called the `createIndex` in the `beforeAll` function.

File: jasmine/spec/inverted-index-test.js (hunk 3 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js]: Thanks. I have added test to verify that the index is correct.

File: jasmine/spec/inverted-index-test.js (hunk 4 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js]: I have added a second index.

I have also added test for when the index has not been uploaded yet. 

Oh, I omitted test for exact result. I will use a much smaller file and add it in the next commit.

File: jasmine/spec/inverted-index-test.js (hunk 5 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js]: Thanks. I have also made corrections for this.

File: jasmine/spec/app-test.js (hunk 6 above)
Comment (desaawa) [reply jasmine/spec/app-test.js]: I need to commit it to the repo or else travis build and code coverage will fail. 

I will try and find an alternative for this. 
Thanks.

File: jasmine/spec/inverted-index-test.js (hunk 4 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js]: Done. I've written test for the exact result of the index.

**Comment by desaawa** [comment]
@q-ode. Please review so I can merge everything and continue my work.
Thanks.

File: jasmine/spec/inverted-index-test.js (hunk 1 above)
Comment (q-ode) [reply jasmine/spec/inverted-index-test.js:37]: I'd rather you created the index here.

Also, the index is meant to be created for just the text. The title if for displaying the result.

File: jasmine/app-test.js [hunk 8]
```diff
@@ -0,0 +1,144 @@
+(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
```
Comment (q-ode) [inline jasmine/app-test.js:1]: I mentioned that generated files shouldn't be committed to the repository.

File: jasmine/spec/inverted-index-test.js [hunk 9]
```diff
@@ -0,0 +1,102 @@
+const books = require('./books.json');
+const books2 = require('./books2.json');
+const books3 = require('./books3.json');
+
+describe('InvertedIndex Class', () => {
+  beforeAll(() => {
+    this.invertedIndex = new InvertedIndex();
+    this.index = this.invertedIndex.createIndex('books', books);
+    this.index2 = this.invertedIndex.createIndex('books2', books2);
+    this.index3 = this.invertedIndex.createIndex('books3', books3);
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js:10]: The inverted index should contain the data from the index is creates. You shouldn't have to assign it to an external variable. I expect that the `getIndex` function would then help you fetch a particular index that has been created.

PS: Except these function are providing some data in the outer scope, please call the methods you want to test in their respective test suites.

File: jasmine/spec/inverted-index-test.js [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+      const index2 = this.invertedIndex.getIndex('books2');
+
+      expect(index1.a).toEqual([0, 1, 2]);
+      expect(index1.alice).toEqual([0]);
+      expect(index2.random).toEqual([1]);
+      expect(index2.room).toEqual([0]);
+
+    });
+
+    it('returns the exact result of the index', () => {
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js:67]: I expect that this satisfies the spec from line 56.

File: jasmine/spec/inverted-index-test.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+    it('should return "undefined" if index does not exist', () => {
+      expect(this.invertedIndex.getIndex(' ')).toEqual(undefined);
+      expect(this.invertedIndex.getIndex('books4')).toEqual(undefined);
+    });
+    
+    it('should return a particular index', () => {
+      const index1 = this.invertedIndex.getIndex('books');
+      const index2 = this.invertedIndex.getIndex('books2');
+
+      expect(index1.a).toEqual([0, 1, 2]);
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js:60]: The spec says `should return a particular index`, I find that lines 57 and 58 have satisfied that.

The next suite then checks that the index returned is accurate. So 60 - 63 seems unnecessary here.

File: jasmine/spec/inverted-index-test.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
+const books = require('./books.json');
+const books2 = require('./books2.json');
+const books3 = require('./books3.json');
+
+describe('InvertedIndex Class', () => {
+  beforeAll(() => {
+    this.invertedIndex = new InvertedIndex();
+    this.index = this.invertedIndex.createIndex('books', books);
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js:8]: Why `this` not `const` or `let`?

File: jasmine/spec/inverted-index-test.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+      expect(this.invertedIndex.getIndex('books2')).toBeTruthy();
+    });
+    it('creates the correct index', () => {
+      expect(this.index.a).toEqual([0, 1, 2]);
+      expect(this.index.alice).toEqual([0]);
+      expect(this.index2.random).toEqual([1]);
+      expect(this.index2.room).toEqual([0]);
+    });
+  });
+
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js:48]: One line spacing is enough. Please apply same to the rest of the file.

File: jasmine/spec/inverted-index-test.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        .toEqual('Index with movies does not exist.');
+    });
+
+    it('should return "not found" for words not in index', () => {
+      expect(this.invertedIndex.searchIndex('', 'books'))
+        .toEqual('no word found');
+    });
+
+    it('should return object with search words', () => {
+      expect(this.invertedIndex.searchIndex('alice unusual wonderland', 'books'))
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js:96]: Lines 96 to 99 are more than 80 characters.

File: jasmine/spec/inverted-index-test.js [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+      .toEqual({'alice': [0], 'unusual': [1,2], 'wonderland': [0]});
+      expect(this.invertedIndex.searchIndex('room', 'books2')).toEqual({'room': [0]});
+      expect(this.invertedIndex.searchIndex('random', 'books2')).toEqual({'random': [1]});
+    });
+  });
+});
```
Comment (q-ode) [inline jasmine/spec/inverted-index-test.js:102]: Great Job here!

I don't see you testing the search through all indexed files though.

File: src/js/inverted-index.js [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+   * @return {Array} array of string tokens
+   */
+  static getWords(text) {
+    return text.toLowerCase().replace(/[^\w\s]/g, '').split(/\s+/);
+  }
+
+  /**
+   * Create index
+   * @function
+   * @param {string} indextitle
```
Comment (q-ode) [inline src/js/inverted-index.js:27]: Index title or filename?

You're creating an index from an uploaded file. I believe that's more accurate.

File: src/js/inverted-index.js [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+   * @param {Array} data
+   * @return {Object} index object
+   */
+  createIndex(indextitle, data) {
+    const dictionary = {};
+
+    data.forEach((doc, arrayIndex) => {
+      const docWords = `${doc.title} + ' ' + ${doc.text}`;
```
Comment (q-ode) [inline src/js/inverted-index.js:35]: You don't need to use the `+` when using string literals.

File: src/js/inverted-index.js [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+ * InvertedIndex class
+ * @class
+ */
+class InvertedIndex {
+  /**
+   * class constructor
+   * @constructor
+   */
+  constructor() {
+    this.indexes = {};
```
Comment (q-ode) [inline src/js/inverted-index.js:11]: If this is supposed to store all the indices created why isn't it declared as an array?

File: src/js/inverted-index.js [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+          }
+        } else {
+          dictionary[word] = [arrayIndex];
+        }
+      });
+    });
+
+    this.indexes[indextitle] = dictionary;
+
+    return this.indexes[indextitle];
```
Comment (q-ode) [inline src/js/inverted-index.js:51]: Why did you need to return the value of `this.indexes[indextitle]` after assigning it in line 49.

I believe it's already available for use from the `getIndex` method.

File: src/js/inverted-index.js [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  /**
+   * Get a particular index
+   * @function
+   * @param {String} indextitle
```
Comment (q-ode) [inline src/js/inverted-index.js:57]: Mentioned this already, it should be filename.

File: src/js/inverted-index.js [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+   */
+  searchIndex(query, indextitle) {
+    const queryWords = InvertedIndex.getWords(query);
+    const index = this.getIndex(indextitle);
+
+    if (!index) {
+      return `Index with ${indextitle} does not exist.`;
+    }
+
+    const result = {};
```
Comment (q-ode) [inline src/js/inverted-index.js:79]: Is `result` an array?

**Review by q-ode — COMMENTED** [review]
Great Job!

Kindly attend to the comments.

**Comment by desaawa** [comment]
Hi @q-ode. I have addressed the issues. You find the new PR [here](https://github.com/andela-aawa/inverted-index/pull/20)

File: src/js/inverted-index.js (hunk 21 above)
Comment (desaawa) [reply src/js/inverted-index.js:79]: I prefer to use an object/hash {}

File: src/js/inverted-index.js (hunk 20 above)
Comment (desaawa) [reply src/js/inverted-index.js:57]: changed.

File: src/js/inverted-index.js (hunk 19 above)
Comment (desaawa) [reply src/js/inverted-index.js:51]: Done.

File: src/js/inverted-index.js (hunk 17 above)
Comment (desaawa) [reply src/js/inverted-index.js:35]: Done

File: src/js/inverted-index.js (hunk 16 above)
Comment (desaawa) [reply src/js/inverted-index.js:27]: Changed.

File: src/js/inverted-index.js (hunk 18 above)
Comment (desaawa) [reply src/js/inverted-index.js:11]: I prefer to use an object/hash {}

File: jasmine/app-test.js (hunk 8 above)
Comment (desaawa) [reply jasmine/app-test.js:1]: I removed it, but my travis build and code coverage all fail as a result

File: jasmine/spec/inverted-index-test.js (hunk 15 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js:102]: I will add the functionality later. For now, I want to make sure the basic search in a particular index works well.

File: jasmine/spec/inverted-index-test.js (hunk 13 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js:48]: Done

File: jasmine/spec/inverted-index-test.js (hunk 9 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js:10]: Done

File: jasmine/spec/inverted-index-test.js (hunk 12 above)
Comment (desaawa) [reply jasmine/spec/inverted-index-test.js:8]: I've changed `this` to `let`
